clamp/clamp.tex

```diff
@@ -0,0 +1,260 @@
+\end{figure}
+
+Since then, the code has been updated to include a more general clampable domain to abstract together parts of the vector and scalar cases:\footnote{As of July 28, 2021.} \url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/clamp.rs#L28-L44} in conjunction with \url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/clamp.rs#L74-L85}. We decided not to refer to the clampable domain in the proofs given that we are separating the vector and scalar cases into two different proof documents.
+
+\begin{figure}[ht]
+    \includegraphics[width=16cm]{clamp3.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+\begin{figure}[ht]
+    \includegraphics[width=16cm]{clamp4.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+\subsection{Pseudocode in Python}\label{sec:pseudocode}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found at \href{https://www.overleaf.com/project/60d215bf90b337ac02200a99}{``List of definitions used in the pseudocode"}. 
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Type \texttt{T} must have trait \texttt{TotalOrd}.\footnote{For now, the OpenDP library only implements \texttt{PartialOrd}, but \texttt{TotalOrd} will soon be implemented. On the architecture meeting of August 4 we discussed different implementations of \texttt{TotalOrd}.}
+    \end{itemize}
+\end{itemize}
+
+\subsubsection*{Postconditions}
+\begin{itemize}
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+\end{itemize}
+
+\begin{lstlisting}[language=Python, escapechar=|] 
+def MakeClamp(L: T, U: T): |\label{line:def}|
+    input_domain = VectorDomain(AllDomain(T))
+    output_domain = VectorDomain(IntervalDomain(L, U)) |\label{line:output}|
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+    
+    def relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+    
+    def function(data: Vec[T]) -> Vec[T]: |\label{line:fn}|
+        def clamp(x: T) -> T: |\label{line:clamp}|
+            return max(min(x, U), L)
+        return list(map(clamp, data)) |\label{line:map}|
+    
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation = relation)
+\end{lstlisting}
+
+
+\section{Proof}
+The necessary definitions for the proof can be found at \href{https://www.overleaf.com/project/60d214e390b337703d200982}{``List of definitions used in the proofs"}.
+
+\subsection{Symmetric Distance}
+\begin{theorem}
+    For every setting of the input parameters \texttt{(L, U)} to \texttt{MakeClamp} such that the given preconditions
+    hold, the transformation returned by \texttt{MakeClamp} has the following properties:
```

Comment:
shouldn't we say "MakeClamp raises a exception (at compile time or run time) or returns a valid transformation."?  (And then we can/should put the def of what constitutes a valid transformation in either the pseudocode_defns or proof_defns file (not sure which).)